### PR TITLE
Add StreamFilterLogs() to enable querying logs by topics

### PR DESCRIPTION
**Pull request by dustinxie**
add a Responder interface to define the actual behavior when a new block arrives, this turns the Listener.Start() into common code. 

the Listener on chain just keep adding Responder -- each implements functionality for different streaming APIs

the original blockListener becomes a Responder, so is log filter added in this PR

### CODE REVIEW

File: api/logfilter.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+				logs = append(logs, log)
+			}
+		}
+	}
+	return logs
+}
+
+// match checks if a given log matches the filter
+func (l *LogFilter) match(log *iotextypes.Log) bool {
+	if l.Address != log.ContractAddress {
```
Comment (zjshen14) [inline api/logfilter.go]: This match should be optional if Address is given

File: api/logfilter.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+		return false
+	}
+	if len(l.Topics) > len(log.Topics) {
```
Comment (zjshen14) [inline api/logfilter.go]: L85-L107 could be simplified into the logic that if there exists at least one topic in log.Topics that doesn't match a one in l.Topics (note that missing isn't considered as mismatch)

File: api/api.go [hunk 3]
```diff
@@ -524,6 +524,56 @@ func (api *Server) StreamBlocks(in *iotexapi.StreamBlocksRequest, stream iotexap
 	}
 }
 
+// StreamFilterLogs streams logs that match the filter condition
+func (api *Server) StreamFilterLogs(in *iotexapi.FilterLogsRequest, stream iotexapi.APIService_StreamFilterLogsServer) error {
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply api/api.go]: fixed in new push

File: api/logfilter.go (hunk 1 above)
Comment (dustinxie) [reply api/logfilter.go]: fixed in new push

**Pull request merged**